### ISSUE TITLE
Drag & Drop target selection highlight wrong target

### ISSUE THREAD

**Issue by elemhsb** [fesch/Structorizer.Desktop#97]
![double_click_failed](https://cloud.githubusercontent.com/assets/2959804/12198348/683c671a-b60f-11e5-8aab-f164cef88782.png)

Picture while dragging LLL to GGG (mouse pointer under (G ) wrong highlight on CCC.

While hunting a strange DOUBLE_CLICK opens randomly a **wrong** other Structure for editing issue -or- move the selectetd structures randomly around, I found that issue and can reproduce it.
It could be a hint to me DOUBLE_CLICK problem. Seen on Windows and Linux >=3.23.08, Mac untested.

**Comment by codemanyak** [comment]
Dragging slowly I can't reproduce it, though while dragging there are always very short flashes on other structures. I think this is due to the recursive selection detection where a first the outer structure feels responsible then asks its substructure if this responds to be selected then the outer structure withdraws its own selection flag. Since drawing is a concurrent thread, I assume some races that are hard to avoid unless the outer structure waits for all substructure before it sets its own selection flag. But in either case, the structure element does not report the selected element to the editor before it's sorted out.
My question would hence be: If you drop the dragged element, does it fall into the place under the cursor or will it be appended to the erroneously green-lighted (outer) element?

**Comment by codemanyak** [comment]
Addendum: That on double clicking a wrong element opens happens rather frequently. I usually make sure to have selected the element to be edited before I double-click, therefore.
I assume races again: The high-prioritized mouse action might interrupt the recursive selection routine prematurely such that the editor still sticks to a cached earlier selection. To avoid this tends to get tricky. It would be about synchronisation of concurrent threads that hardly know of each other at application programming level and were breakpoints are of little or no help at all.

**Comment by codemanyak** [comment]
Anyway, I wonder how you caught this screenshot. Some high-resolution screen motion recording tool?

**Comment by elemhsb** [comment]
The element falls correct to the current mouse position.
Highlight the next outer structure also cause when hitting a vertical line. 

I am glad to here that the double clicking problem is known. ;-)
A workaround is to close structorizer end reopen again. Sometimes i get java exceptions and continue work, i should write a issue.

No special screen capture tool, i only pressed PRINT for screen capture under Debian / Cinnamon Desktop.

**Comment by codemanyak** [comment]
Okay, under Linux I am able to reproduce the effect, too. It's stable enough. (Under Windows 7 it is so volatile and fugitive - just flashes of some milliseconds - that I could either take snapshots of it nor had I had an idea how to analyse it. The reason might be that the default mouse event frequency is apparently much higher under Windows.) So, obviously, races are not to be blamed but tiny gaps between the regions claimed by the substructure. I'll try to enlarge the claimed area by 1 pixel. I think, a little overlapping would be less uncomfortable because it's just the first who claims the cursor will win it.

**Comment by codemanyak** [comment]
With respect to the exceptions, I found a line in the code where a null pointer exception could have been provoked in rare cases. The next version won't contain this trap anymore.

**Comment by codemanyak** [comment]
Now I am beginning to run out of ideas - even an overlapping of the claims by more than 3 pixels does not eliminate the complained behaviour. Maybe it's yet races. I'll try to change the flagging order.

**Comment by elemhsb** [comment]
Today (now) i ***reuse*** an opened structorizer 3.23-07 on MacOS from yesterday evening which loaded test structures. yesterday the double click error didn't occur!
Now it does .... software degrade over time 8-(.
Selection (on click) sometime 60% open direct the edit dialog. double click opens a lot of edit edit structures in series. Double click interval is middle.

**Comment by fesch** [comment]
I can reproduce it on my Mac if the mouse is on the line between DDD and EEE.
A similar problem occurs if the mouse is between EEE and GGG (two elements
![structorizer - double_click_failed nsd](https://cloud.githubusercontent.com/assets/1027354/12215843/2fefe54e-b6cd-11e5-9bd3-2df8e1955be8.jpg)
 highlighted)

**Comment by codemanyak** [comment]
@fesch What you caught is the case that the entire subqueue of the false branch of the outer alternative (CCC) claims the selection while both its children (EEE, GGG) don't do so.
I seems that modifying the coordinate comparison in Element.getElementByCoord() such that equality is involved slightly improves the situation though it doesn't eliminate the issue. It' seems just a little bit more fugitive. Overdoing it (i.e. allowing overlapping) makes things worse again, it seems. This has some logic because the substructure recursion is always done completely (against my first assumption), which is necessary to unselect all elements that are not (but may have been before) beneath the cursor. Hence overlapping may provoke unwanted multiple selection (though it does not explain why the parent structure still claims the selection. I am really puzzled. Will need to do a complete logging with all rectangles and coordinates involved to get clue what the hell is going on here.

**Comment by codemanyak** [comment]
I just tried a very old version: 3.19. Back then the selection on dragging still worked steadily, without flickering, though when the cursor stood over a separating line, then the parent structure got selected instead of some its substructure elements as well. So the gap problem had always existed.
The important difference however, seems to be that in these early versions the element to be moved had not been hovering transparently over the diagram as is does now. (Instead, just the selected and the target cell got green or red, and on releasing the cursor in green state, the flopped into its new place.
Hence, it might be the dynamic drawing of the hovering copy that makes things a little wobbly. And I remember it dating back some years that I first observed this flickering and unsteady highlighting after something with the dragging had changed. Now I see more clearly, it might have been the hovering - my guess is, that it might irritate the selection process.

**Comment by codemanyak** [comment]
Eventually found some time to proceed with the analysis. My logging showed that the faulty selection is caused by some - still unidentified - event that recomputes all element rectangles with upper left corner at the coordinates origin, i. e. all start at (0,0). That's why substructures (like EEE or (GGG) in the example) temporarily don't feel responsible at the cursor position, and the selection falls back to a larger enclosing element (CCC) which rather accidently still contains the cursor coordinate.
The Arranger wasn't open (a drawing update in the Arranger definitely re-computes element bounds with Arranger coordinates in order to draw the shared diagrams at their respective position on the Arranger surface, but usually the bounds are re-computed again by Structorizer afterwards).
I am still working on the identification of the causing misguiding event or a change of the drawing mechanism towards a position-independent bounds caching (which might also save time because relative bounds need re-computing only after some diagram or representation change).

**Comment by codemanyak** [comment]
So, the issue is as clear as daylight now: It is indeed a concurrency problem. The element identification for a given cursor position id based on the rect attribute, but this attribute is set by two methods in a different way. First method prepareDraw() assigns mere extensions (upper left being 0, 0, see previous comment). Then method draw() replaces these rectangles by located ones, i.e. representing the real coordinates of the drawn element. Element selection and identification only work with rectangles of the latter kind.
Since the "mouse-dragged" events are asynchronous, they may occur between the prepareDraw() and the draw() method invocation in which case the rect values are useless and produce wrong target selection. Letting method prepareDraw() work with a local variable or a new separate attribute (e. g. rect0) seems to fix the problem pretty fine at first glance.
But then some new trouble becomes obvious:

1. On dragging a structured element over one of its substructure elements the "blushing" in red wouldn't work anymore. This is due to the fact that the transparently hovering element isn't a clone but the element itself drawn a second time, now at the cursor position (which affects rect). Though the original rect value is restored afterwards, this doesn't hold for the element's substructure elements where the displaced coordinates remain.
2. With a clone of the diagram placed in the Arranger, on the other hand, the position of the transparently hovering element isn't where it belongs but relates to the corresponding Arranger surface position instead. This is because the Arranger drawing update will be done after the Structorizer drawing, such that the rect found on dragging may be the Arranger   

So, now we know what's going on, but how to solve it? The best thing might be to store only position-independent data in the elements, i.e. origin-bound rectangles (or just a pair of width and height, to eliminate redundancy) and to provide the translation coordinate as method argument. Structured elements should the also store the relative upper-left corner coordinates of their substructure (like loop bodies, branches etc.) The draw method already carries a position argument (_upper_left) ofg which only the left and top components are used. This argument could easily be adapted on recursion for the proposed new mechanism. Conversely, the cursor position could be recursively relativated on element identification/selection. In cases of direct access to a substructure element this could obtain the diagram top left position and thus its own displacement by an ascending recursion. This paradigm change might even reduce the event handling time. Elements' rect (and substructure location) attributes could last till the next editing or display setting change,

**Comment by codemanyak** [comment]
The issue is fixed now, eventually. Bugfix3.23 branch holds the bugfix. It follows the oulined plan and definitely works under Windows and Linux (openSuSe). All that flickering is gone. A new Structorizer version 3.23-15 may be now created via the pending pull request.

**Comment by codemanyak** [comment]
@elemhsb: New version is available on the Structorizer home page.

**Comment by codemanyak** [comment]
An update to the drawing reorganisation initiated by this bugfix was necessary in order to invalidate the cached drawing infos of all diagrams parked in the Arranger on changes to global apparence settings. (Available on bugfix3.23 branch, to be integrated in the next version, certainly 3.23-17)

**Comment by elemhsb** [comment]
@codemanyak Year, you got it! My self compiled version 3.24-02 works very well!
Structorizer is back to be a real productivity tool. Thanks a lot for your patience.
